### PR TITLE
refactoring

### LINKED ISSUE #53: MSAL loginRedirect does not work properly.
If I use the MSAL loginRedirect method and if I run the sample and use email and password to login it does not login the user or call the handle callback method. Also, clicking on Google button to sigin with Google does not work.

### PR DESCRIPTION
**Pull request by derisen**
# CHANGELOG

- Source code refactored.
- Ajax calls replaced with Fetch API.
- Corrected HTML structure.
- UI redesigned with Bootstrap 4 framework.
- Corrected README.md.
- Added login with redirect (as an optional script file).
- Authentication decoupled from access token request

PR for the [B2C docs](https://docs.microsoft.com/en-us/azure/active-directory-b2c/tutorial-single-page-app?tabs=applications) covering this sample is [here](https://github.com/MicrosoftDocs/azure-docs-pr/pull/105885)

### CODE REVIEW

File: CONTRIBUTING.md
Comment (sameerag) [inline CONTRIBUTING.md]: Do we need a single command that does linting / testing and starts the project for anyone who wants to contribute?

Comment (derisen) [reply CONTRIBUTING.md]: actually we have no tests or linting script atm. Suppose I should just remove this part?

Comment (sameerag) [reply CONTRIBUTING.md]: In that case please remove.

File: JavaScriptSPA/api.js
Comment (sameerag) [inline JavaScriptSPA/api.js]: If we are getting thew token here and also accessing the resource, how is it resource_access code only? Can we have some glue code that passes the acquired token to the api.js?

File: JavaScriptSPA/apiConfig.js
Comment (sameerag) [inline JavaScriptSPA/apiConfig.js]: Why is request here if the file is named `apiConfig.js`?

File: JavaScriptSPA/authConfig.js
Comment (sameerag) [inline JavaScriptSPA/authConfig.js:20]: Why is request here if the file is named `authConfig.js`?

File: JavaScriptSPA/authRedirect.js
Comment (sameerag) [inline JavaScriptSPA/authRedirect.js]: Lets sync offline on this, we do this check in the library too and looks redundant for me. But I do understand the value in adding it here. Helps to spend some time to establish a `redirect` API usage pattern for `msal js`. I can back port this to dev samples too if needed.

File: JavaScriptSPA/policies.js
Comment (sameerag) [inline JavaScriptSPA/policies.js]: Are we planning to add different policies later?

File: JavaScriptSPA/apiConfig.js
Comment (derisen) [reply JavaScriptSPA/apiConfig.js]: wouldn't it make sense to group the api configuration with the token scopes it accepts? would you prefer to move it into api.js instead?

File: JavaScriptSPA/authConfig.js
Comment (derisen) [reply JavaScriptSPA/authConfig.js:20]: same as above. move to auth.js instead?

File: JavaScriptSPA/policies.js
Comment (derisen) [reply JavaScriptSPA/policies.js]: might be a good way to organize strings if the user wants to try out different policies to compare. what would you suggest?

File: JavaScriptSPA/authRedirect.js
Comment (derisen) [reply JavaScriptSPA/authRedirect.js]: you're probably right. I removed this and so far there are no issues -took this bit uncritically from dev samples.

File: JavaScriptSPA/api.js
Comment (derisen) [reply JavaScriptSPA/api.js]: makes sense. actually moving this bit to authPopup.js and its redirect version to authRedirect.js made it more convenient. Now the user just have to switch the files to switch the flows.

File: JavaScriptSPA/authRedirect.js
Comment (sameerag) [inline JavaScriptSPA/authRedirect.js:76]: So here is the thing ... @mmacy had the feedback that API access code needs to be separated from token acquisition! By having `callApi()` in the same file as `getToken` we are defeating that purpose. Is there a way we can pass tokens from a file to the other and restrict `callApi()` to API access only? We can do that in index.html also.

Same applies for VanillaJS samples (non B2C I mean).

cc @mmacy to add his feedback.

Comment (mmacy) [reply JavaScriptSPA/authRedirect.js:76]: @sameerag @derisen I haven't yet had a chance to run through the tutorials with the new sample, but essentially, the user needs to be able to:

1. Tutorial 1: Sign in to the application
1. Tutorial 2: Access the API

At the completion of tutorial 1, the user won't yet have setup or configured the API pieces, so sign-in needs to work on its own. Then, at the end of tutorial 2, after they've set up the API pieces, they hit the "access API" button and that's when it calls the API.

File: JavaScriptSPA/policies.js
Comment (mmacy) [reply JavaScriptSPA/policies.js]: Requiring the user to edit another file to get it working with their tenant adds to the complexity of steps in the tutorial.

Having constants in another object is certainly something that might be done in a real-world app, but in a day-1 learning situation, it's better to present more literal code. In this case, I'd like to keep the URI with the policy name intact as opposed to concatenation.

In these types of tutorials, to quote Dave Cheney, [clear is better than clever](https://dave.cheney.net/2019/07/09/clear-is-better-than-clever).

Comment (derisen) [reply JavaScriptSPA/policies.js]: sounds good @mmacy

File: JavaScriptSPA/authRedirect.js
Comment (derisen) [reply JavaScriptSPA/authRedirect.js:76]: Popup flow works just fine without anything api related. I've moved `callAPI()` (it's badly named, all it does is to get an access token and pass it to the `callApiWithAccessToken()` function) back to `api.js`, then simply didn't include those api related files in the loadup, and it authenticates all fine (`authPopup.js` just gets an id token). So there should be no problem for tutorials that way.

The issue is with the redirect flow. I call `callApiWithAccessToken()` in the redirect callback function, so when the page loads with the access token it calls the API.
 
To fully seperate these logics I can't think of a way right now other than having a global for accessToken in auth.js files. Then we would just pass it in the html like: 

`<button onclick="callApi(accessToken)">Call API</button>`

I guess it would be best to figure this out after we sync @sameerag ?

Comment (mmacy) [reply JavaScriptSPA/authRedirect.js:76]: From the tutorial flow standpoint, as long as it doesn't throw an error after sign-in but before they've finished tutorial 2 and set up the API bits, we're golden.

**Comment by mmacy** [comment]
~~Hi @sameerag & @derisen, how we looking on this one? Has the code stabilized such that I'm good to go on updating the corresponding docs in https://github.com/MicrosoftDocs/azure-docs-pr/pull/105885 to work with the code in this PR?~~

PR status is being updated in https://github.com/Azure-Samples/active-directory-b2c-javascript-msal-singlepageapp/pull/52#pullrequestreview-366205296

Cc: @msmimart

File: JavaScriptSPA/authRedirect.js
Comment (sameerag) [reply JavaScriptSPA/authRedirect.js:76]: @derisen Apologies on the delay on sync up. Does today work for you? Please message me. @mmacy we will have an answer to you by tomorrow on the status.

Comment (derisen) [reply JavaScriptSPA/authRedirect.js:76]: @sameerag no worries! sure thing, let me know time

Comment (mmacy) [reply JavaScriptSPA/authRedirect.js:76]: @sameerag @derisen No worries, and no hurry - I have *plenty* on my plate. :wink:

Comment (sameerag) [reply JavaScriptSPA/authRedirect.js:76]: @mmacy an update: We talked (and a few more msal js team members) and except finalizing the `redirect` code,  we have reached to a conclusion on most questions. We will try to close the pending AI next week. Thanks.

**Comment by mmacy** [comment]
@derisen @sameerag @negoe Another doc that will need to be updated to coincide with the merging of this PR:

https://docs.microsoft.com/azure/active-directory/develop/msal-b2c-overview

File: JavaScriptSPA/authConfig.js
Comment (sameerag) [inline JavaScriptSPA/authConfig.js]: Couple things: 
- msal-js configuration is here: `https://azuread.github.io/microsoft-authentication-library-for-js/docs/msal/modules/_configuration_.html`
- msal-js request(AuthneticationParameters) is here: `https://azuread.github.io/microsoft-authentication-library-for-js/docs/msal/modules/_authenticationparameters_.html`

Please add both in the comments and teh respctive content

File: JavaScriptSPA/authConfig.js
Comment (sameerag) [inline JavaScriptSPA/authConfig.js]: Change the comment to "Please set this true to save cache in cookies to address trusted zones limitations in IE" - and add the corresponding documentation in MSAL wiki.

Comment (derisen) [reply JavaScriptSPA/authConfig.js]: I've set the default value to false as well (that's the recommendation, right?)

File: JavaScriptSPA/authRedirect.js
Comment (derisen) [reply JavaScriptSPA/authRedirect.js:76]: @mmacy @sameerag I think we are good to go now. From the tutorials perspective, here's what I think:

The default flow for the tutorial should be Popup. We also provide an optional `authRedirect.js` file for people who want try it, but it should be tried after the tutorials. 

For part 2: "Grant access to API", the user will do these:
- create `api.js` file to call the api
- create `apiConfig.js` file to enter the api parameters
- go back to `authPopup.js` and add the function `passTokenToApi()`
- go back to `authConfig.js` file to add `tokenRequest` scopes
- go back to `index.html` file to to add the "Call API" button

Even using `authRedirect.js` will not break anything, they will still login all fine, its just that the code has more references to `callApiWithAccessToken` function due to the nature of redirect flow (so its more tedious to go back and add this line where its needed), but since there'll be no "Call API" button on the user interface, it will not run (and even if it runs, nothing will break, just a white error).

Comment (sameerag) [reply JavaScriptSPA/authRedirect.js:76]: Sounds good me. Excellent improvement overall! cc @ktsakas @jmprieur

**Review by sameerag — APPROVED** [review]
LGTM and approving. `Redirect` code can be improved and we are doing some upgrades. Will reach out once `msal-core` is Ready for the same.

**Comment by derisen** [comment]
Thank you so much @sameerag ! Merging this now... (? @mmacy)

**Comment by mmacy** [comment]
@derisen Please do not merge yet. Now that this sample has solidified, I can continue and complete the updates to the articles that reference it in this PR:

https://github.com/MicrosoftDocs/azure-docs-pr/pull/105885

Might be a few days, but I'll shoot for getting it merged and published by end of the week (Friday, ~~20~~ 27 March 2020). I'll get in touch with you via Teams to coordinate the merge/publish so that it all lands together.

**Comment by derisen** [comment]
no worries, i got my lesson last time : ) @mmacy

**Comment by sameerag** [comment]
@derisen I would also suggest for any merge to the master branch in any sample, please get approvals from people verbally in the github. That is usually the norm :-)

**Comment by derisen** [comment]
mea culpa : ) @sameerag

**Comment by mmacy** [comment]
@derisen Hey Dogen, does this PR fix https://github.com/MicrosoftDocs/azure-docs/issues/47225 by chance? That is, does it demonstrate the ability to switch authority to support multiple user flows/policies?

**Comment by derisen** [comment]
@mmacy It doesn't : ( Support for multiple policies seems to be in the [todo stage](https://github.com/AzureAD/microsoft-authentication-library-for-js/projects/1#card-24423974), though there is a [workaround](https://github.com/Azure-Samples/active-directory-b2c-javascript-msal-singlepageapp/issues/9). @sameerag to confirm

**Comment by sameerag** [comment]
@mmacy unfortunately we do not yet support multiple accounts, there is on going work for 'msal-node` to support this which will be back ported to `msal-core` in future; The timelines aren't very clear at this point though.

cc @DarylThayil

File: JavaScriptSPA/authConfig.js
Comment (mmacy) [inline JavaScriptSPA/authConfig.js]: @derisen  Does this const still need to be defined here if it's already _apiConfig.js_?

From _apiConfig.js_ L2-4:

```JavaScript
// The current application coordinates were pre-registered in a B2C tenant.
const apiConfig = {
  b2cScopes: ["https://fabrikamb2c.onmicrosoft.com/helloapi/demo.read"],
  webApi: "https://fabrikamb2chello.azurewebsites.net/hello"
```

Comment (derisen) [reply JavaScriptSPA/authConfig.js]: technically it isn't. I wanted to make `authConfig.js` file self-contained, and use `apiConfig.js` file only to clarify and make explicit what are the relation of these strings to the API.

Comment (derisen) [reply JavaScriptSPA/authConfig.js]: Do you think this is more clear?:

```
const tokenRequest = {
  scopes: apiConfig.b2cScopes 
}
```

Comment (mmacy) [reply JavaScriptSPA/authConfig.js]: Well, I'm unsure whether it's more clear, but it seems odd to define the same value in two different consts in the same project. I haven't tested, but would another option be to swap out `tokenRequest` from _authPopup.js_ and _authRedirect.js_ and instead reference `apiconfig.b2cScopes`, something like:

```JavaScript
// Acquires and access token and then passes it to the API call
function passTokenToApi() {
  getTokenPopup(apiConfig.b2cScopes)
    .then(tokenResponse => {
```

Then you could remove the API config entirely from _authConfig.js_, and it'd only be in _apiConfig.js_.

Comment (mmacy) [reply JavaScriptSPA/authConfig.js]: Actually, that wouldn't work since `getTokenPopup` is probably expecting a `scopes` property on the passed-in object.

Comment (derisen) [reply JavaScriptSPA/authConfig.js]: how about?

```
const tokenRequest = {
  scopes: apiConfig.b2cScopes  //e.g. ["https://fabrikamb2c.onmicrosoft.com/helloapi/demo.read"]
}
```

Comment (mmacy) [reply JavaScriptSPA/authConfig.js]: I like it. The developer needs then to update the API scope URI in only one const in the project, but they're also made aware of the type it's expecting.

Comment (derisen) [reply JavaScriptSPA/authConfig.js]: alright pushing the changes

Comment (mmacy) [reply JavaScriptSPA/authConfig.js]: @sameerag Are you OK with this change? It avoids having to specify the same const value twice, while still leaving `tokenRequest` available in _authConfig.js_.

Comment (derisen) [reply JavaScriptSPA/authConfig.js]: Can we get this going today @sameerag @mmacy ? Hopefully it will help with a user's issue : ))

Comment (mmacy) [reply JavaScriptSPA/authConfig.js]: Super close - I just need someone to walk through all the docs as if they've never seen them before to make sure all the steps function as expected and result in a successful experience. I've done so, but I'd like at least one other person to help verify.

Here are the staging links:

* [Quickstart](https://review.docs.microsoft.com/en-us/azure/active-directory-b2c/quickstart-single-page-app?branch=pr-en-us-105885)
* [Tutorial - SPA](https://review.docs.microsoft.com/en-us/azure/active-directory-b2c/tutorial-single-page-app?branch=pr-en-us-105885&tabs=app-reg-preview) (part 1)
* [Tutorial - SPA calls API](https://review.docs.microsoft.com/en-us/azure/active-directory-b2c/tutorial-single-page-app-webapi?branch=pr-en-us-105885&tabs=app-reg-preview) (part 2)

Corresponding PR in docs: https://github.com/MicrosoftDocs/azure-docs-pr/pull/105885

Comment (sameerag) [reply JavaScriptSPA/authConfig.js]: > @sameerag Are you OK with this change? It avoids having to specify the same const value twice, while still leaving `tokenRequest` available in _authConfig.js_.

Yes @mmacy. 
``` javascript
const tokenRequest = {
  scopes: apiConfig.b2cScopes 
} 
```
looks like proper approach, as we do not want to pass `scopes` as an individual object.

**Pull request merged**
